From 25e2c72de2de2edda8d78dceda81c03bcc91abdb Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Tue, 13 Apr 2021 09:37:53 -0500
Subject: [PATCH 079/196] Make non-pushonly scriptsig in coinbase inputs
 standard (#144)

Currently, non-pushonly scriptSigs are non-standard; but mining pool software relies on the scriptSig of the coinbase being arbitray.
We enforce standardness on all txs, so blocks having non-pushonly coinbases scriptSigs will be rejected, which is not what we want.

This will remove the "scriptsig-not-pushonly" restriction for standardness on coinbase inputs.
---
 src/policy/policy.cpp          | 3 ++-
 src/test/transaction_tests.cpp | 8 ++++++++
 2 files changed, 10 insertions(+), 1 deletion(-)

diff --git a/src/policy/policy.cpp b/src/policy/policy.cpp
index c8beed772..80edcdcda 100644
--- a/src/policy/policy.cpp
+++ b/src/policy/policy.cpp
@@ -88,7 +88,8 @@ bool IsStandardTx(const CTransaction &tx, bool permit_bare_multisig,
             reason = "scriptsig-size";
             return false;
         }
-        if (!txin.scriptSig.IsPushOnly()) {
+        // Coinbase can have non-pushonly scriptSig
+        if (!tx.IsCoinBase() && !txin.scriptSig.IsPushOnly()) {
             reason = "scriptsig-not-pushonly";
             return false;
         }
diff --git a/src/test/transaction_tests.cpp b/src/test/transaction_tests.cpp
index e3d5be2f5..a6cb59433 100644
--- a/src/test/transaction_tests.cpp
+++ b/src/test/transaction_tests.cpp
@@ -919,9 +919,17 @@ BOOST_AUTO_TEST_CASE(test_IsStandard) {
         uint8_t orig_op = *prev_pc;
         // replace current push-op with each non-push-op
         for (auto op : non_push_ops) {
+            // save prevout
+            COutPoint prevout = t.vin[0].prevout;
             t.vin[0].scriptSig[index] = op;
             BOOST_CHECK(!IsStandardTx(CTransaction(t), reason));
             BOOST_CHECK_EQUAL(reason, "scriptsig-not-pushonly");
+
+            // make input coinbase, then scriptsig-not-pushonly does not apply
+            t.vin[0].prevout = COutPoint();
+            BOOST_CHECK(IsStandardTx(CTransaction(t), reason));
+            // restore prevout
+            t.vin[0].prevout = prevout;
         }
         // restore op
         t.vin[0].scriptSig[index] = orig_op;
-- 
2.39.5 (Apple Git-154)

